### PR TITLE
Tpd 1041/business entity functionality

### PR DESCRIPTION
**Pull request by josearmandohf**
Adding Business record type logic while keeping Person record type functionality stable.
[touchpointsoftware.atlassian.net/browse/TPD-1041](https://touchpointsoftware.atlassian.net/browse/TPD-1041)

### CODE REVIEW

File: CmsWeb/Areas/People/Views/Person/Index.cshtml [hunk 1]
```diff
@@ -6,19 +6,28 @@
...
     {
         showResourcesTab = Model.PeopleId == ViewExtensions2.CurrentDatabase.UserPeopleId || User.IsInRole("ViewResources");
     }
+    showResourcesTab = !businessrecordtype ? showResourcesTab : false;
 
     var showMinistryTab = Util.IsMyDataUser ? false : !RoleChecker.HasSetting(SettingName.HideMinistryTab, false);
+    showMinistryTab = !businessrecordtype ? showMinistryTab : false;
+
+    var showInvolvementTab = !businessrecordtype ? true : false;
+    var showProfileTab = !businessrecordtype ? true : false;
```
Comment (brondavies) [inline CmsWeb/Areas/People/Views/Person/Index.cshtml:29]: `? true : false` is redundant here.

File: CmsWeb/Content/touchpoint/js/people/person.js [hunk 2]
```diff
@@ -95,6 +95,25 @@
...
+    $('body').on('click', '#convertRecordType', function (ev) {
+        ev.preventDefault();
+        var href = $(this).attr("href");
+
+        swal({
+            title: "Are you sure?",
+            type: "warning",
+            showCancelButton: true,
+            confirmButtonClass: "btn-danger",
+            confirmButtonText: "Yes, convert it!",
```
Comment (brondavies) [inline CmsWeb/Content/touchpoint/js/people/person.js:107]: remove the `!`

File: CmsData/QueryBuilder/FieldMap.csv [hunk 3]
```diff
@@ -3,6 +3,7 @@ Group,Group,,,,Grouping,,,"Groups conditions or other groups together in an ""al
 Age,NullInteger,,,,Personal,,,Enter an age. Leave blank to find those without a DOB.
 GenderId,Code,,GenderCodes,,Personal,Gender,,Select Gender from the list.
 MaritalStatusId,Code,,MaritalStatusCodes,,Personal,Marital Status,,Select Marital Status from the list.
+PeopleRecordType,Code,,PeopleRecordCodes,,Personal,People Record Type,,Select Record Type from the list.
```
[the remark this answers is not in the document]
Comment (josearmandohf) [reply CmsData/QueryBuilder/FieldMap.csv]: I've change it to RecordType

File: CmsData/Generated/Family.cs [hunk 4]
```diff
@@ -300,8 +305,25 @@ public bool RecordStatus
 
 		}
 
-		
-		[Column(Name="BadAddressFlag", UpdateCheck=UpdateCheck.Never, Storage="_BadAddressFlag", DbType="bit")]
+        [Column(Name = "IsBusiness", UpdateCheck = UpdateCheck.Never, Storage = "_IsBusiness", DbType = "bit")]
+        public bool? IsBusiness
```
[the remark this answers is not in the document]
Comment (josearmandohf) [reply CmsData/Generated/Family.cs]: This is done

File: CmsData/QueryBuilder/Expressions/Personal.cs [hunk 5]
```diff
@@ -110,5 +110,26 @@ internal Expression HasPicture()
                 expr = Expression.Not(expr);
             return expr;
         }
+        internal Expression PeopleRecordType()
```
[the remark this answers is not in the document]
Comment (josearmandohf) [reply CmsData/QueryBuilder/Expressions/Personal.cs]: Done

File: CmsWeb/Areas/People/Views/Person/Index.cshtml (hunk 1 above)
Comment (josearmandohf) [reply CmsWeb/Areas/People/Views/Person/Index.cshtml]: Fixed

File: CmsWeb/Content/touchpoint/js/people/person.js (hunk 2 above)
Comment (josearmandohf) [reply CmsWeb/Content/touchpoint/js/people/person.js]: Done

File: CmsData/QueryBuilder/Expressions/Personal.cs [hunk 6]
```diff
@@ -110,5 +110,26 @@ internal Expression HasPicture()
                 expr = Expression.Not(expr);
             return expr;
         }
+        internal Expression RecordType()
+        {
+            var codes0 = CodeValues.Split(',').Select(ff => ff.Trim()).ToList();
```
Comment (jason-rice) [inline CmsData/QueryBuilder/Expressions/Personal.cs:115]: can you rename your variables to something more descriptive? for example, instead of codes0, you could name it codeList. this makes the code more readable and easier to understand. Otherwise I have to remember what code0, q, and any other variables actually mean instead of just reading the variable name.

File: CmsWeb/Areas/Search/Models/PeopleSearchModel.cs [hunk 7]
```diff
@@ -259,6 +260,19 @@ where p.CellPhone.Contains(m.communication)
...
+            
+            switch (m.recordtype)
+            {
+                case 0:
+                    people = people.Where(p => p.IsBusiness == false || p.IsBusiness == null);
+                    break;
+                case 1:
+                    people = people.Where(p => p.IsBusiness == true);
+                    break;
+                //case 2:
```
Comment (jason-rice) [inline CmsWeb/Areas/Search/Models/PeopleSearchModel.cs:272]: is there a reason you left this commented out code in the PR?

File: CmsWeb/Areas/Search/Models/SearchAdd/PendingPersonModel.cs [hunk 8]
```diff
@@ -161,8 +197,14 @@ where pids.Contains(p.PeopleId)
         }
 
         public bool IsNewFamily { get; set; }
```
Comment (jason-rice) [inline CmsWeb/Areas/Search/Models/SearchAdd/PendingPersonModel.cs:199]: It would be better if all of the properties of this class were up at the top of class, near line 27. That would make the code easier read and comprehend, and maintain in the future.

File: CmsWeb/Areas/Search/Models/SearchAdd/SearchAddModel.cs [hunk 9]
```diff
@@ -89,7 +90,11 @@ public SearchAddModel(CMSDataContext db, string context, string contextid, bool
             }
         }
 
-        public List<PendingPersonModel> PendingList { get; set; }
+        public List<PendingPersonModel> PendingList
```
Comment (jason-rice) [inline CmsWeb/Areas/Search/Models/SearchAdd/SearchAddModel.cs:93]: what's the reason for this change?

File: CmsData/Extensions/Person.cs [hunk 10]
```diff
@@ -2161,7 +2161,18 @@ public static int FetchOrCreateVolunteerApplicationStatus(CMSDataContext db, str
             }
             return i.Id;
         }
-
+        public static int FetchOrCreateVolunteerCode(CMSDataContext db, string code)
```
Comment (brondavies) [inline CmsData/Extensions/Person.cs:2164]: This doesn't look like it's used.  Is this just for data conversion?

File: CmsData/Generated/Family.cs [hunk 11]
```diff
@@ -109,8 +111,11 @@ public partial class Family : INotifyPropertyChanging, INotifyPropertyChanged
 		
 		partial void OnCreatedDateChanging(DateTime? value);
 		partial void OnCreatedDateChanged();
-		
-		partial void OnRecordStatusChanging(bool value);
+
+        partial void OnIsBusinessChanging(bool? value);
```
Comment (brondavies) [inline CmsData/Generated/Family.cs:115]: should match the type for `bool IsBusiness` (not nullable)

File: CmsData/Generated/Person.cs [hunk 12]
```diff
@@ -630,6 +633,9 @@ public partial class Person : INotifyPropertyChanging, INotifyPropertyChanged
         partial void OnMemberAnyChurchChanging(bool? value);
         partial void OnMemberAnyChurchChanged();
 
+        partial void OnIsBusinessChanging(bool? value);
```
Comment (brondavies) [inline CmsData/Generated/Person.cs:636]: Same here

File: CmsData/QueryBuilder/Expressions/Personal.cs (hunk 6 above)
Comment (josearmandohf) [reply CmsData/QueryBuilder/Expressions/Personal.cs:115]: Sure, I'll do it.

File: CmsWeb/Areas/Search/Models/PeopleSearchModel.cs (hunk 7 above)
Comment (josearmandohf) [reply CmsWeb/Areas/Search/Models/PeopleSearchModel.cs:272]: Nope, it was only redundant code. I'll remove it.

File: CmsWeb/Areas/People/Models/Person/PersonModel.cs [hunk 13]
```diff
@@ -154,6 +158,12 @@ public AddressInfo PrimaryAddr
             }
         }
 
+        public bool IsBusiness
```
Comment (brondavies) [inline CmsWeb/Areas/People/Models/Person/PersonModel.cs:161]: Can be simplified to `public bool IsBusiness { get; set; }` and remove `private bool? isBusiness;`

File: CmsWeb/Areas/Search/Models/SearchAdd/SearchAddModel.cs (hunk 9 above)
Comment (josearmandohf) [reply CmsWeb/Areas/Search/Models/SearchAdd/SearchAddModel.cs:93]: I was only to debug getter and setter but I'll put it back the way it was.

File: CmsData/Extensions/Person.cs (hunk 10 above)
Comment (josearmandohf) [reply CmsData/Extensions/Person.cs:2164]: Yes, this was one change made on provisioning system. I can take it off this PR.

File: CmsWeb/Areas/People/Views/Person/Personal/Header.cshtml [hunk 14]
```diff
@@ -3,16 +3,27 @@
...
-</h1>
+    <h1 id="nameline" class="@(Model.Person.IsDeceased == true ? " alert alert-danger" : "" )">
+        @if (Model.IsBusiness)
+        {
+        @Model.LastName
+        }
+        else
+        {
+        @Model.Name
+        }
```
Comment (brondavies) [inline CmsWeb/Areas/People/Views/Person/Personal/Header.cshtml:14]: This could be simpler if `PersonModel` implemented `GetName()` with this logic

File: CmsData/Generated/Family.cs (hunk 11 above)
Comment (josearmandohf) [reply CmsData/Generated/Family.cs]: No, it shouldn't. I'll fix it.

File: CmsWeb/Areas/Search/Models/SearchAdd/PendingPersonModel.cs [hunk 15]
```diff
@@ -19,24 +19,43 @@ public class PendingPersonModel : IDbBinder
...
         {
-            CurrentDatabase = db;
+            CurrentDatabase = db;            
         }
 
         public CMSDataContext CurrentDatabase { get; set; }
         public int index { get; set; }
         public string context { get; set; }
-
+        private string _FirstName;
```
Comment (brondavies) [inline CmsWeb/Areas/Search/Models/SearchAdd/PendingPersonModel.cs:36]: I don't think you need this

File: CmsWeb/Areas/Search/Models/SearchAdd/PendingPersonModel.cs [hunk 16]
```diff
@@ -19,24 +19,43 @@ public class PendingPersonModel : IDbBinder
...
+            }
+        }        
 
         [StringLength(25)]
         public string NickName { get; set; }
 
         [StringLength(25)]
         public string MiddleName { get; set; }
-
+        private string _LastName;
```
Comment (brondavies) [inline CmsWeb/Areas/Search/Models/SearchAdd/PendingPersonModel.cs:53]: don't need this

File: CmsWeb/Areas/Search/Models/SearchAdd/PendingPersonModel.cs [hunk 17]
```diff
@@ -161,8 +197,14 @@ where pids.Contains(p.PeopleId)
         }
 
         public bool IsNewFamily { get; set; }
+        private bool _IsBusiness;
+        [DisplayName("Add a business/entity")]
+        public bool IsBusiness {
+            get { return _IsBusiness; }
+            set { _IsBusiness = value; }
+        }
```
Comment (brondavies) [inline CmsWeb/Areas/Search/Models/SearchAdd/PendingPersonModel.cs]: You can just use a simple property `public bool IsBusiness { get; set; }`

File: CmsWeb/Areas/Search/Models/SearchAdd/SearchAddModel.cs [hunk 18]
```diff
@@ -166,20 +172,34 @@ public int NextNewFamilyId()
...
+            else
+            {
+                p.FamilyId = familyid;
+                p.index = PendingList.Count;
+                p.Gender = new CodeInfo(99, "Gender");
+                p.MaritalStatus = new CodeInfo(99, "MaritalStatus");
+                p.Campus = new CodeInfo(CampusId, campuslist);
+                p.EntryPoint = new CodeInfo(EntryPointId, "EntryPoint");
+                p.context = AddContext;
+                p.IsBusiness = isbusiness;
```
Comment (brondavies) [inline CmsWeb/Areas/Search/Models/SearchAdd/SearchAddModel.cs]: you should move the common lines of code between these blocks before the `if` and simplify

File: CmsWeb/Areas/People/Views/Person/Personal/Header.cshtml (hunk 14 above)
Comment (josearmandohf) [reply CmsWeb/Areas/People/Views/Person/Personal/Header.cshtml:14]: Yeah, that is a good one, thanks!

File: CmsWeb/Areas/Search/Models/SearchAdd/SearchAddModel.cs (hunk 18 above)
Comment (josearmandohf) [reply CmsWeb/Areas/Search/Models/SearchAdd/SearchAddModel.cs]: Right... on it.
